src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp

```diff
@@ -154,33 +154,59 @@ class LoaderTreeNode : public ResourceObj {
   // one.
   int _num_folded;
 
-  void print_with_childs(outputStream* st, BranchTracker& branchtracker,
-      bool print_classes, bool verbose) const {
+  // Returns Klass of loader; NULL for bootstrap loader
+  const Klass* loader_klass() const {
+    return (_loader_oop != NULL) ? _loader_oop->klass() : NULL;
+  }
 
-    ResourceMark rm;
+  // Returns ResourceArea-allocated class name of loader class
+  const char* loader_class_name() const {
+    const Klass* klass = loader_klass();
+    return klass != NULL ? klass->external_name() : NULL;
+  }
 
-    if (_cld == NULL) {
-      // Not sure how this could happen: we added a preliminary node for a parent but then never encountered
-      // its CLD?
-      return;
+  // Returns oop of loader name; NULL for bootstrap; NULL if no name was set
+  oop loader_name_oop() const {
+    return (_loader_oop != NULL) ? java_lang_ClassLoader::name(_loader_oop) : NULL;
+  }
+
+  // Returns ResourceArea-allocated name of loader, "" if none is set
+  const char* loader_name() const {
+    oop name_oop = loader_name_oop();
+    return name_oop != NULL ? java_lang_String::as_utf8_string(name_oop) : "";
+  }
+
+  bool is_bootstrap() const {
+    if (_loader_oop == NULL) {
+      assert(_cld != NULL && _cld->is_boot_class_loader_data(), "bootstrap loader must have CLD");
+      return true;
     }
+    return false;
+  }
+
+  void print_with_child_nodes(outputStream* st, BranchTracker& branchtracker,
+      bool print_classes, bool verbose) const {
+
+    assert(SafepointSynchronize::is_at_safepoint(), "invariant");
+
+    ResourceMark rm;
 
     // Retrieve information.
-    const Klass* const loader_klass = _cld->class_loader_klass();
-    const Symbol* const loader_name = _cld->name();
+    const Klass* const the_loader_klass = loader_klass();
+    const char* const the_loader_class_name = loader_class_name();
+    const char* const the_loader_name = loader_name();
 
     branchtracker.print(st);
 
     // e.g. +-- "app", jdk.internal.loader.ClassLoaders$AppClassLoader
     st->print("+%.*s", BranchTracker::twig_len, "----------");
-    if (_cld->is_the_null_class_loader_data()) {
+    if (is_bootstrap()) {
       st->print(" <bootstrap>");
     } else {
-      assert(!_cld->has_class_mirror_holder(), "_cld must be the primary cld");
-      if (loader_name != NULL) {
-        st->print(" \"%s\",", loader_name->as_C_string());
+      if (the_loader_name[0] != '\0') {
+        st->print(" \"%s\",", the_loader_name);
       }
-      st->print(" %s", loader_klass != NULL ? loader_klass->external_name() : "??");
+      st->print(" %s", the_loader_class_name);
```

Comment:
`the_loader_class_name` could be NULL here. You need to change `loader_class_name()` t return "??" like the original code uses.

Reply:
Good catch. We never encounter this because Klass can only be null for the bootstrap loader, which is handled in a different branch. I'll fix it.